forem/forem
# Add review item model and mark as reviewed for mod center

## What type of PR is this? (check all applicable)

## Description
This is a proof of concept of the "mark as reviewed" functionality for mods to review articles in a more inbox-style way through the Mod Center. The general idea is this:

1. an article is created, and is put into a review queue, via a `ReviewItem` creation
2. the Mod Center shows articles that need to be reviewed
3. a moderator visits the mod center, and takes an action, which marks an article as reviewed

We're still finalizing the details of what exactly a reviewed article means. This current implementation is based on these ideas:
- the Mod Center is showing **only articles you have not reviewed**
- an article that has been reviewed by someone will still be unreviewed to other moderators

I was having trouble figuring out the best way to implement this. Currently, each article will create a `ReviewItem`. This allows us to easily query for which articles to display:
```ruby
article_ids = ReviewItem.where(reviewed: false, reviewable_type: "Article").pluck(:reviewable_id)

articles = Article.where(id: article_ids)
```
However, this means that we'll have **at least** one new row per article. Additional rows per article would be linked to a `reviewer`. Like so:
```ruby
ReviewItem.find_by(reviewable: some_article, reviewer_id: nil, reviewed: false)
ReviewItem.find_by(reviewable: some_article, reviewer_id: 11, reviewed: true)
ReviewItem.find_by(reviewable: some_article, reviewer_id: 12, reviewed: true)
# ... and so on
```
I feel like this could be problematic, such as always having a `ReviewItem` with `reviewer_id: nil` and will never become `reviewed: true`.

Another implementation I thought of was making the `reviewer_id` column an array column:
```ruby
# on article creation
ReviewItem.find_by(reviewable: some_article, reviewer_ids: [], reviewed: false)
# on article being reviewed by someone
ReviewItem.update(reviewable: some_article, reviewer_ids: [1], reviewed: true)
# querying for whether current_user has reviewed it
current_user_id = 1
ReviewItem.find_by(reviewable: some_article, reviewer_ids: [1])
# actually not sure how ActiveRecord works with Postgres array column yet, but it can be solved!
```

**_Anyway, additional thoughts will definitely be appreciated!!!_**

## QA Instructions, Screenshots, Recordings

![Screen Shot 2020-06-30 at 11 09 00 AM](https://user-images.githubusercontent.com/17884966/86143190-21b2a900-bac2-11ea-9601-7a3d859b6293.png)

Steps for QA:
1. Create review items in console:
```
articles = Article.last(3)
articles.each do |article|
  ReviewItem.create!(reviewable: article, reviewer_id: nil)
end
2. Go to the Mod Center `/mod`
3. Open an article, and take any action on it via the mod panel
4. Refresh the page to see that it's marked as reviewed (yes, this is kind of janky right now but it's just a proof of concept)

## Added tests?
- [ ] no, not yet

## Added to documentation?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)
